What is NumPy?

NumPy (Numerical Python) is a Python library used for:

Fast mathematical calculations
Working with arrays
Linear algebra
Statistics
Machine Learning
Deep Learning

In [4]:
import numpy as np 
numbers=np.array([1,2,3,4])
print(numbers*2)

[2 4 6 8]


Level 2: Creating Arrays

In [5]:
import numpy as np 
arr=np.array([1,2,3,4])
print(arr)

[1 2 3 4]


In [6]:
arr=np.array([
    [1,2,3],
    [4,5,6]
])
print(arr)

[[1 2 3]
 [4 5 6]]


In [7]:
# 3d array 
arr=np.array([
    [
        [1,2],
        [3,4]
    ],
    [
        [5,6],
        [7,8]
    ]
])
print(arr)


[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]


In [8]:
#array attribute
arr = np.array([[1,2,3],[4,5,6]])

In [10]:
#shape 
print(arr.shape)

(2, 3)


In [11]:
#dimenions 
print(arr.ndim)

2


In [12]:
#size: counts tota elements
print(arr.size)

6


In [13]:
#data types 
print(arr.dtype)

int64


In [14]:
# creating special arryas 
#zeros
np.zeros((3,4))

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [15]:
np.ones((2,3))

array([[1., 1., 1.],
       [1., 1., 1.]])

In [ ]:
#identity matrix(eye) : used to make vector 
np.eye(5)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

In [17]:
#random numbers 
np.random.rand(3)

array([0.29489245, 0.16491914, 0.05352726])

In [18]:
#random integers 
np.random.randint(1,10,5)

array([9, 3, 5, 6, 4], dtype=int32)

In [19]:
#array indexing 
arr=np.array([10,20,30,40])

In [20]:
#first element 
arr[0]

np.int64(10)

In [21]:
arr[-1]#last elememt 

np.int64(40)

In [22]:
 #for 2D
arr=np.array([
    [1,2,3],
    [4,5,6]
])

In [23]:
#get 4
arr[1,0]
#get 5
arr[1,1]

np.int64(5)

In [24]:
# slicing 
arr=np.array([10,20,30,40,50])
arr[1:4]

array([20, 30, 40])

slicing vs indexing : indexing gives exact number , slicing : slices 

```
testing maarkdown
a=np.array([1,2,3])
b=np.array([4,5,6])
print(a+b)
```

In [38]:
#indexing 
arr=np.array([20,30,40,50])
arr[2]

np.int64(40)

In [25]:
arr = np.array([
    [1, 2, 3],
    [4, 5, 6]
])
print(arr[:, 1])

[2 5]


Mathematical Operations

In [27]:
a=np.array([1,2,3])
b=np.array([4,5,6])
print(a+b)

[5 7 9]


In [29]:
print (a-b)

[-3 -3 -3]


In [31]:
print (a*b)

[ 4 10 18]


In [32]:
# power 
a**2

array([1, 4, 9])

In [33]:
np.sqrt(a)

array([1.        , 1.41421356, 1.73205081])

In [ ]:
#broadcasting : in NumPy is a mechanism that allows arithmetic operations between arrays of different shapes.
arr=np.array([1,2,3,4,5])
arr+5

array([ 6,  7,  8,  9, 10])

In [35]:
matrix=np.array([
    [1,2,3],
    [4,5,6]
])
matrix+np.array([10,20,30])

array([[11, 22, 33],
       [14, 25, 36]])

In [40]:
#reshapinfg 
arr=np.arange(12)
print(arr)

[ 0  1  2  3  4  5  6  7  8  9 10 11]


In [41]:
# convert into 3 rows and 4 ols 
arr.reshape(3,4)

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [ ]:
# flatten : converts matrix into singe row 
arr.flatten()

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [44]:
#statistics
np.mean(arr)

np.float64(5.5)

In [49]:
np.max(arr)

np.int64(11)

In [50]:
np.min(arr)

np.int64(0)

In [51]:
np.median(arr)

np.float64(5.5)

In [52]:
np.std(arr)

np.float64(3.452052529534663)

In [ ]:
#variance : a statistical measurement that shows how far a set of numbers are spread out from their average (mean) value. 

np.var(arr)

np.float64(11.916666666666666)

Linear Algebra

@ operator is the matrix multiplication operator.

In [ ]:
A=np.array([
    [1,2],
    [3,4]
])
B=np.array([
    [5,6],
    [7,8]
])
A@B



array([[19, 22],
       [43, 50]])

In [55]:
#Matrix × Vector
A=np.array([
    [1,2],
    [3,4]
])
v=np.array([10,20])
print(A@v)

[ 50 110]


In [ ]:
#Vector X matrix  (dot product)
v=np.array([10,20])
A=np.array([
    [1,2],
    [3,4]
])
print(v@A)

[ 70 100]


In [57]:
#2*3 matrix nd 3*2 matrix 
A=np.array([
    [1,2,3],
    [4,5,6]
])
B=np.array([
    [7,8],
    [9,10],
    [11,12]
])
print(A@B)

[[ 58  64]
 [139 154]]


Transpose:

 Transpose :eans swapping the rows and columns of a matrix.
 if matrix has:(rows, columns) then converting it to (columns , rows)



In [58]:
A=np.array([
    [1,2,3],
    [4,5,6]
])
print(A)

[[1 2 3]
 [4 5 6]]


In [ ]:
# using transpose 
print(A.T) # swaps rows and columns 

[[1 4]
 [2 5]
 [3 6]]


In [61]:
A=np.array([
    [1,2],
    [3,4]

])
print(A.T)

[[1 3]
 [2 4]]


#Inverse:the inverse of a matrix is similar to the reciprocal of a number

In [62]:
np.linalg.inv(A)

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [ ]:
A=np.array([
    [3,4],
    [5,6]
])
np.linalg.inv(A) #linalg stands for linear algebra

array([[-3. ,  2. ],
       [ 2.5, -1.5]])

Determinant:A determinant is a single number that tells you whether a matrix can be reversed (has an inverse) and how it changes the size of shapes.
##Determinant ≠ 0 → The matrix has an inverse and can be "undone."
##Determinant = 0 → The matrix has no inverse because it squashes space (like a square into a line).

In [67]:
import numpy as np 
A=np.array([
    [1,2],
    [3,4]
])
det=np.linalg.det(A)
print(det)

-2.0000000000000004


In [68]:
A_inv=np.linalg.inv(A)
print(A_inv)

[[-2.   1. ]
 [ 1.5 -0.5]]


In [70]:
# matrix with determinant =0
import numpy as np

A = np.array([
    [1, 2],
    [2, 4]
])

det = np.linalg.det(A)

print(det)

0.0


In [ ]:
# finding its reverse 
np.linalg.inv(A) # this fails beacuse the determinant is 0

LinAlgError: Singular matrix

det != 0 → ✅ Inverse exists.
det == 0 → ❌ No inverse exists.

Eigenvalues :Eigenvalues are numbers that show how much a matrix stretches, shrinks, or flips its special vectors (eigenvectors) without changing their directio

In [72]:
import numpy as np 
A=np.array([
    [2,0],
    [0,3]

])
print(A)

[[2 0]
 [0 3]]


In [ ]:
# finding eigen values and eigen vectors 
eigenvalues, eigenvectors = np.linalg.eig(A)

print("Eigenvalues:")
print(eigenvalues)

print("Eigenvectors:")
print(eigenvectors)

Eigenvalues:
[2.+0.j 3.+0.j]
Eigenvectors:
[[1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j]]


Boolean Masking/filter:Boolean masking means selecting data from an array using conditions that return True or False.

In [75]:
import numpy as np 
arr=np.array([1,5,8,10,3,7])
mask=arr>5
print(mask)

[False False  True  True False  True]


In [76]:
import numpy as np 
arr= np.array([20,30,40,50,60,70,80,90])
mask=arr>40
print(mask)

[False False False  True  True  True  True  True]


Vectorization:means performing operations on entire arrays at once instead of looping through each element one by one in Python.

In [3]:
import numpy as np
arr=np.array([1,2,3,4,5])
result=arr+10
print(result)

[11 12 13 14 15]


In [5]:
square=arr**2
print(square)

[ 1  4  9 16 25]


In [8]:
arr=np.array([4,5,6,7])
arr =arr+5
print(arr)

[ 9 10 11 12]


In [9]:
arr=np.array([1,4,9,16])
print(np.sqrt(arr))

[1. 2. 3. 4.]


Random Module 


In [15]:
np.random.seed(42)
print(np.random.rand())
print(np.random.rand())
print(np.random.rand())
#rand() generates a random floating-point number between
# 0 (inclusive) and 1 (exclusive).

0.3745401188473625
0.9507143064099162
0.7319939418114051


In [ ]:
np.random.rand(5) #random decimal

array([0.59865848, 0.15601864, 0.15599452, 0.05808361, 0.86617615])

What is a Normal Distribution?

A normal distribution (also called a Gaussian distribution) is a bell-shaped distribution where:

Mean (μ) = 0
Standard Deviation (σ) = 1

-1≤Z≤1

In [20]:
#random normal distribution 
#np.random: numpy random module 
#.randn: random number from normal distribution ,gnerates 5 normal numbers
np.random.randn(5)

array([-0.2257763 ,  0.0675282 , -1.42474819, -0.54438272,  0.11092259])

ouutput 
-0.2257763	Slightly below the mean (0)
0.0675282	Very close to the mean
-1.42474819	About 1.42 standard deviations below the mean
-0.54438272	Below the mean
0.11092259	Slightly above the mean

In [21]:

np.random.seed(42)

print(np.random.randn(5))

[ 0.49671415 -0.1382643   0.64768854  1.52302986 -0.23415337]


In [22]:
np.random.seed(42)

print(np.random.randn(5))

[ 0.49671415 -0.1382643   0.64768854  1.52302986 -0.23415337]


Shuffle: randomy rearranges the elements of an array in place

In [23]:
#shuffle 
np.random.shuffle(arr)

In [25]:
#1D Array 
import numpy as np

arr = np.array([1, 2, 3, 4, 5])

np.random.shuffle(arr)

print(arr)

[1 3 4 2 5]


In [27]:
arr=np.array([
    [1,2],
    [3,4],
    [5,6]
])
np.random.shuffle(arr)
print(arr)

[[1 2]
 [5 6]
 [3 4]]


Permutation: used to create a randomly permuted(shuffled) copy of sequence or array 
Unlike shuffle() it doesnot moddify the original array

syntax 
np.random.permuttion(x)

In [28]:
import numpy as np 
result =np.random.permutation(5)
print(result )

[2 1 4 0 3]


Permutation and Combination are counting techniques used in mathematics and probability.

The easiest way to remember them is:

Permutation = Order matters
Combination = Order doesn't matter

In [29]:
import numpy as np

arr = np.array([1,2,3,4])

print(np.random.permutation(arr))

[3 1 4 2]


In [32]:
import numpy as np 
from itertools import combinations
arr=np.array([1,2,3,4,5])
result=list(combinations(arr,2))
print(result)

[(np.int64(1), np.int64(2)), (np.int64(1), np.int64(3)), (np.int64(1), np.int64(4)), (np.int64(1), np.int64(5)), (np.int64(2), np.int64(3)), (np.int64(2), np.int64(4)), (np.int64(2), np.int64(5)), (np.int64(3), np.int64(4)), (np.int64(3), np.int64(5)), (np.int64(4), np.int64(5))]


In [ ]:
import numpy as np
from itertools import combinations
 
arr = np.array([1,2,3,4])

result = list(combinations(arr, 3))

print(result)

[(np.int64(1), np.int64(2), np.int64(3)), (np.int64(1), np.int64(2), np.int64(4)), (np.int64(1), np.int64(3), np.int64(4)), (np.int64(2), np.int64(3), np.int64(4))]


In [37]:
np.random.choice([1,2,3],size=5)

array([1, 1, 3, 3, 3])